main.py

```diff
@@ -81,12 +81,33 @@ def swap_track(track_id):
     return [11, 14, 9, 16, 15, 10, 13, 12][track_id - 9]
 
 
+def tail_call_gen(func: typing.Callable[[...], typing.Generator]):
+    """
+    Decorator to implement tail call optimisation as a generator
+
+    instead of calling itself the tail of the function should
+    yield the arguments for its call. the facilitator can then
+    handle the calls in a way far more lightweight than the call
+    stack, and avoiding blowing the stack with really deep recursion.
+    """
+    @functools.wraps(func)
+    def facilitator(*args):
+        argslist = [args]
+
+        while argslist:
+            args = [*func(*argslist.pop())]
+            argslist.extend(reversed(args))
+
+    return facilitator
+
+
 # main generation function. uses given board and variables to generate tracks every game movement.
 # VERY condensed: it first looks if cars are on switches and does interaction shenanigans, then
 # generates tracks for every car BASED ON ITS VELOCITY. it then does checks for the generated tracks to determine
 # which ones are valid, and adds those as well as the generated car for each track to a list.
 # after every car is generated, it gets every possible track combination between every car, and calls the function
 # again with every new game movement.
+@tail_call_gen
```

Comment:
If you've not seen this syntax before, this is a decorator, and it's just syntactic sugar where

```python
@decorator
def function(*args):
    ...
```
is equivalent to

```python
def function(*args):
    ...
function = decorator(function)
```